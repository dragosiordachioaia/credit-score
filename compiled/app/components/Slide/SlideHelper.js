"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getDebtSlide = exports.getBalanceSlide = exports.getOffersSlide = exports.getScoreSlide = void 0;

var _react = _interopRequireDefault(require("react"));

var _bem = _interopRequireDefault(require("../../utils/bem"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var cn = (0, _bem.default)({
  block: 'slide'
});

var getScoreSlide = function getScoreSlide(props, state) {
  var currentSlideData = props.slides[state.currentSlideIndex];
  var crtScoreToDisplay = state.currentCoefficientNumber * currentSlideData.maxScore;
  return _react.default.createElement("div", {
    key: "slide-score",
    className: cn('individual-slide'),
    style: {
      width: "".concat(props.radius * 2, "px")
    }
  }, _react.default.createElement("p", {
    className: cn('small-text')
  }, "Your credit score is"), _react.default.createElement("h2", {
    className: cn('score-text'),
    style: {
      color: props.slides[state.currentSlideIndex].color
    }
  }, Math.round(crtScoreToDisplay)), _react.default.createElement("p", {
    className: cn('small-text')
  }, "out of", _react.default.createElement("b", {
    className: cn('max-score')
  }, state.maxScore)), _react.default.createElement("p", {
    className: cn('description-text'),
    style: {
      color: props.slides[state.currentSlideIndex].color
    }
  }, "Soaring high"));
};

exports.getScoreSlide = getScoreSlide;

var getOffersSlide = function getOffersSlide(props, state) {
  var currentSlideData = props.slides[state.currentSlideIndex];
  var crtScoreToDisplay = state.currentCoefficientNumber * currentSlideData.maxScore;
  return _react.default.createElement("div", {
    key: "slide-offers",
    className: cn('individual-slide'),
    style: {
      width: "".concat(props.radius * 2, "px")
    }
  }, _react.default.createElement("h2", {
    className: cn('score-text'),
    style: {
      color: props.slides[state.currentSlideIndex].color
    }
  }, Math.round(crtScoreToDisplay)), _react.default.createElement("p", {
    className: cn('small-text')
  }, "New offers"));
};

exports.getOffersSlide = getOffersSlide;

var getBalanceSlide = function getBalanceSlide(props, state) {
  return _react.default.createElement("div", {
    key: "slide-offers",
    className: cn('individual-slide'),
    style: {
      width: "".concat(props.radius * 2, "px")
    }
  }, _react.default.createElement("img", {
    src: "/cards.png",
    className: cn('cards')
  }), _react.default.createElement("p", {
    className: cn('small-text')
  }, "Transfer your ", _react.default.createElement("br", null), " balance!"));
};

exports.getBalanceSlide = getBalanceSlide;

var getDebtSlide = function getDebtSlide(props, state) {
  var currentSlideData = props.slides[state.currentSlideIndex];
  var crtScoreToDisplay = state.currentCoefficientNumber * currentSlideData.maxScore;
  var changeText = "\xA3".concat(Math.abs(currentSlideData.change));

  if (currentSlideData.change < 0) {
    changeText = "-".concat(changeText);
  }

  return _react.default.createElement("div", {
    key: "slide-debt",
    className: cn('individual-slide'),
    style: {
      width: "".concat(props.radius * 2, "px")
    }
  }, _react.default.createElement("p", {
    className: cn('small-text')
  }, "Your long term debt is"), _react.default.createElement("h2", {
    className: cn('score-text'),
    style: {
      color: props.slides[state.currentSlideIndex].color
    }
  }, "\xA3", Math.round(crtScoreToDisplay)), _react.default.createElement("p", {
    className: cn('small-text')
  }, "Change since last check: ", _react.default.createElement("b", null, changeText)), _react.default.createElement("p", {
    className: cn('description-text'),
    style: {
      color: props.slides[state.currentSlideIndex].color
    }
  }, "You're doing great!"));
};

exports.getDebtSlide = getDebtSlide;