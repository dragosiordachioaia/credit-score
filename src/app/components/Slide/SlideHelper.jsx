import React from 'react';
import bemHelper from '../../utils/bem';

const cn = bemHelper({ block: 'slide' });

export const getScoreSlide = (props, state) => {
  const currentSlideData = props.slides[state.currentSlideIndex];
  const crtScoreToDisplay = state.currentCoefficientNumber * currentSlideData.maxScore;

  return (
    <div
      key='slide-score'
      className={cn('individual-slide')}
      style={{width: `${props.radius * 2}px`}}
    >
      <p className={cn('small-text')}>
        Your credit score is
      </p>
      <h2
        className={cn('score-text')}
        style={{color: props.slides[state.currentSlideIndex].color}}
      >
        {Math.round(crtScoreToDisplay)}
      </h2>
      <p className={cn('small-text')}>
        out of
        <b className={cn('max-score')}>{state.maxScore}</b>
      </p>
      <p
        className={cn('description-text')}
        style={{color: props.slides[state.currentSlideIndex].color}}
      >
        Soaring high
      </p>
    </div>
  )
}

export const getOffersSlide = (props, state) => {
  const currentSlideData = props.slides[state.currentSlideIndex];
  const crtScoreToDisplay = state.currentCoefficientNumber * currentSlideData.maxScore;

  return (
    <div
      key='slide-offers'
      className={cn('individual-slide')}
      style={{width: `${props.radius * 2}px`}}
    >
      <h2
        className={cn('score-text')}
        style={{color: props.slides[state.currentSlideIndex].color}}
      >
        {Math.round(crtScoreToDisplay)}
      </h2>
      <p className={cn('small-text')}>
        New offers
      </p>
    </div>
  )
}

export const getBalanceSlide = (props, state) => {
  return (
    <div
      key='slide-offers'
      className={cn('individual-slide')}
      style={{width: `${props.radius * 2}px`}}
    >
      <img src='/cards.png' className={cn('cards')}></img>
      <p className={cn('small-text')}>
        Transfer your <br /> balance!
      </p>
    </div>
  )
}

export const getDebtSlide = (props, state) => {
  const currentSlideData = props.slides[state.currentSlideIndex];
  const crtScoreToDisplay = state.currentCoefficientNumber *currentSlideData.maxScore;

  let changeText = `£${Math.abs(currentSlideData.change)}`;
  if(currentSlideData.change < 0) {
    changeText = `-${changeText}`;
  }

  return (
    <div
      key='slide-debt'
      className={cn('individual-slide')}
      style={{width: `${props.radius * 2}px`}}
    >
      <p className={cn('small-text')}>
        Your long term debt is
      </p>
      <h2
        className={cn('score-text')}
        style={{color: props.slides[state.currentSlideIndex].color}}>
        £{Math.round(crtScoreToDisplay)}
      </h2>
      <p className={cn('small-text')}>
        Change since last check: <b>{changeText}</b>
      </p>
      <p
        className={cn('description-text')}
        style={{color: props.slides[state.currentSlideIndex].color}}
      >
        You're doing great!
      </p>
    </div>
  )
}
