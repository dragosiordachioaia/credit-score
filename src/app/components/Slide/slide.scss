$border-width: 1px;
$slide-radius: 150px;

@keyframes scale-in{
  0% {
    clip-path: circle(0px at center);
  }
  100% {
    clip-path: circle($slide-radius at center);
  }
}

.slide {
  &--main {
    position: absolute;
    top: calc(50vh - #{$slide-radius});
    left: calc(50vw - #{$slide-radius});
    width: $slide-radius * 2;
    height: $slide-radius * 2;
    border-radius: 50%;
    clip-path: circle(0px at center);

    animation-name: scale-in;
    animation-duration: 0.7s;
    animation-iteration-count: 1;
    animation-delay: 0s;
    animation-fill-mode: forwards;
  }
  &__bg {
    z-index: -1;
    width: 100vw;
    height: 100vh;
    top: 0px;
    left: 0px;
    position: fixed;
    background-color: lightblue;
    background: no-repeat top;
    background-size: 100vw 100vh;
    background-image: url('/_background--dashboard.png');
    filter: blur(12px) brightness(1.1);
  }
  &__border {
    width: calc(100% - #{$border-width*2});
    height: calc(100% - #{$border-width * 2});
    border-radius: 50%;
    border: $border-width solid rgba(255, 255, 255, 0.25);
    position: absolute;
    top: 0;
    left: 0;
  }
  &__content {
    width: calc(100% - #{$border-width*2});
    height: calc(100% - #{$border-width * 2});
    position: absolute;
    top: 20%;
    left: 0;
    text-align: center;
    color: #fff;
  }
  &__reel {
    transition: all 400ms;
    position: absolute;
    top: 0;
  }
  &__score-text {
    font-size: 4em;
    margin-top: 0.2em;
  }
  &__max-score {
    margin-left: 5px;
  }
  &__individual-slide {

  }
  &__description-text {
    margin-top: 1.5em;
  }
}
