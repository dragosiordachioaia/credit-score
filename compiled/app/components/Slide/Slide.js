"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _bem = _interopRequireDefault(require("../../utils/bem"));

var _TweenMax = _interopRequireDefault(require("gsap/TweenMax"));

var _EasePack = _interopRequireDefault(require("gsap/EasePack"));

var _Constants = require("../../constants/Constants");

var _SlideHelper = require("./SlideHelper");

var _underscore = _interopRequireDefault(require("underscore"));

require("./slide.scss");

var _SVGArc = _interopRequireDefault(require("../SVGArc/SVGArc"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } _setPrototypeOf(subClass.prototype, superClass && superClass.prototype); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.getPrototypeOf || function _getPrototypeOf(o) { return o.__proto__; }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

var cn = (0, _bem.default)({
  block: 'slide'
});

var Slide =
/*#__PURE__*/
function (_Component) {
  function Slide(props) {
    var _this;

    _classCallCheck(this, Slide);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Slide).call(this, props));
    _this.state = {
      // the value that we want to display in the end, after the animation has ended
      targetScore: 0,
      // the maximum value that is possible to display, used to compute the coefficients
      maxScore: 0,
      // float between 0 and 1, to be used for showing the animated circular path
      currentCoefficientStroke: 0,
      // float between 0 and 1, to be used for displaying the big score text
      currentCoefficientNumber: 0,
      // the index of the current slide
      currentSlideIndex: 0
    };
    _this.animationDuration = 2; // these are the intermediate values for the tweens; we declare them here
    // just to know about them

    _this.valueTweenScore = 0;
    _this.valueTweenStroke = 0; // we need to keep a reference to the tween objects so we can kill them when unmounting

    _this.tweenStroke = null;
    _this.tweenScore = null;
    _this.startAnimation = _this.startAnimation.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.displayContentReel = _this.displayContentReel.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.displayMarkers = _this.displayMarkers.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.killTweens = _this.killTweens.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.onSlideClick = _this.onSlideClick.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.onNewProps = _this.onNewProps.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    return _this;
  }

  _createClass(Slide, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.onNewProps();
    }
  }, {
    key: "onNewProps",
    value: function onNewProps() {
      var newState = {
        targetScore: this.props.slides[0].score,
        maxScore: this.props.slides[0].maxScore,
        currentSlideIndex: 0
      };
      var stateChangeCallback = this.startAnimation;
      this.killTweens();

      if (!this.props.animate) {
        newState.currentCoefficientStroke = this.props.slides[0].score / this.props.slides[0].maxScore;
        newState.currentCoefficientNumber = newState.currentCoefficientStroke;
        stateChangeCallback = undefined;
      }

      this.setState(newState, stateChangeCallback);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.killTweens();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      // this should probable be done in a smarter way
      if (JSON.stringify(this.props.slides) !== JSON.stringify(prevProps.slides)) {
        this.onNewProps();
      }
    }
  }, {
    key: "startAnimation",
    value: function startAnimation() {
      var _this2 = this;

      this.killTweens(); // we reset this so it always grows from 0 to our value for the desired effect
      // we do not neet to reset the 'stroke' one, because we want that one to
      // animate from the current position to the next

      this.valueTweenScore = 0;
      this.tweenScore = _TweenMax.default.to(this, this.animationDuration, {
        valueTweenScore: this.state.targetScore / this.state.maxScore,
        ease: _EasePack.default.Strong.easeOut,
        // we use the update handle from this tween to update both values instead
        // of doing it for both tweens in order to avoid re-rendering twice per animation cycle
        onUpdate: function onUpdate() {
          _this2.setState({
            currentCoefficientNumber: _this2.valueTweenScore,
            currentCoefficientStroke: _this2.valueTweenStroke
          });
        }
      });
      this.tweenStroke = _TweenMax.default.to(this, this.animationDuration, {
        valueTweenStroke: this.state.targetScore / this.state.maxScore,
        ease: _EasePack.default.Bounce.easeOut
      });
    }
  }, {
    key: "killTweens",
    value: function killTweens() {
      if (this.tweenStroke) {
        this.tweenStroke.kill();
      }

      if (this.tweenScore) {
        this.tweenScore.kill();
      }
    }
  }, {
    key: "displayContentReel",
    value: function displayContentReel() {
      var _this3 = this;

      return this.props.slides.map(function (slideData) {
        switch (slideData.type) {
          case 'score':
            return (0, _SlideHelper.getScoreSlide)(_this3.props, _this3.state);

          case 'offers':
            return (0, _SlideHelper.getOffersSlide)(_this3.props, _this3.state);

          case 'debt':
            return (0, _SlideHelper.getDebtSlide)(_this3.props, _this3.state);

          case 'balance':
            return (0, _SlideHelper.getBalanceSlide)(_this3.props, _this3.state);

          default:
            return null;
        }
      });
    }
  }, {
    key: "getRadius",
    value: function getRadius(size) {
      var radius = _Constants.slideSizes.big;

      if (_Constants.slideSizes.hasOwnProperty(size)) {
        radius = _Constants.slideSizes[size];
      }

      return radius;
    }
  }, {
    key: "onSlideClick",
    value: function onSlideClick() {
      var nextSlide = this.state.currentSlideIndex + 1;

      if (this.state.currentSlideIndex >= this.props.slides.length - 1) {
        nextSlide = 0;
      }

      var stateChangeCallback;

      if (this.props.animate) {
        stateChangeCallback = this.startAnimation;
      }

      this.setState({
        currentSlideIndex: nextSlide,
        maxScore: this.props.slides[nextSlide].maxScore,
        targetScore: this.props.slides[nextSlide].score
      }, stateChangeCallback);
    }
  }, {
    key: "displayMarkers",
    value: function displayMarkers() {
      var _this4 = this;

      if (this.props.slides.length <= 1) {
        return null;
      }

      return this.props.slides.map(function (slideData, index) {
        var className = cn('marker');

        if (index === _this4.state.currentSlideIndex) {
          className += " " + cn('marker-selected');
        }

        return _react.default.createElement("div", {
          key: "slide-".concat(index),
          className: className
        });
      });
    }
  }, {
    key: "render",
    value: function render() {
      if (!this.state.currentCoefficientStroke || !this.state.currentCoefficientNumber || !this.state.maxScore) {
        return null;
      }

      var radius = this.getRadius(this.props.size);
      var strokeWidth = this.props.strokeWidth || 3;
      var spaceToEdge = 4;
      var angle = this.state.currentCoefficientStroke * 360;

      if (angle % 360 === 0) {
        angle = 359.9;
      }

      return _react.default.createElement("div", {
        className: cn(null, "main-".concat(this.props.size)),
        style: this.props.style,
        onClick: this.onSlideClick
      }, _react.default.createElement("div", {
        className: cn('bg')
      }), _react.default.createElement("div", {
        className: cn('border')
      }), _react.default.createElement(_SVGArc.default, {
        strokeWidth: strokeWidth,
        radius: radius - strokeWidth - spaceToEdge,
        angle: angle,
        color: this.props.slides[this.state.currentSlideIndex].color
      }), _react.default.createElement("div", {
        className: cn('content')
      }, _react.default.createElement("div", {
        className: cn('reel'),
        style: {
          left: "-".concat(this.state.currentSlideIndex * radius * 2, "px"),
          width: "".concat(this.props.slides.length * radius * 2, "px")
        }
      }, this.displayContentReel())), _react.default.createElement("div", {
        className: cn('marker-container')
      }, this.displayMarkers()));
    }
  }]);

  _inherits(Slide, _Component);

  return Slide;
}(_react.Component);

exports.default = Slide;
module.exports = exports["default"];