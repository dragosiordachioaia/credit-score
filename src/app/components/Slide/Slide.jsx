import React, { Component } from 'react';
import bemHelper from '../../utils/bem';
import TweenMax from 'gsap/TweenMax';
import EaseTypes from 'gsap/EasePack';

import { slideSizes } from '../../constants/Constants';

import {
  getScoreSlide,
  getOffersSlide,
  getBalanceSlide,
  getDebtSlide,
} from './SlideHelper';
import _ from 'underscore';

import './slide.scss';

const cn = bemHelper({ block: 'slide' });

import SVGArc from '../SVGArc/SVGArc';

export default class Slide extends Component {
  constructor(props) {
    super(props);
    this.state = {
      // the value that we want to display in the end, after the animation has ended
      targetScore: 0,

      // the maximum value that is possible to display, used to compute the coefficients
      maxScore: 0,

      // float between 0 and 1, to be used for showing the animated circular path
      currentCoefficientStroke: 0,

      // float between 0 and 1, to be used for displaying the big score text
      currentCoefficientNumber: 0,

      // the index of the current slide
      currentSlideIndex: 0,
    };
    this.animationDuration = 2;

    // these are the intermediate values for the tweens; we declare them here
    // just to know about them
    this.valueTweenScore = 0;
    this.valueTweenStroke = 0;

    // we need to keep a reference to the tween objects so we can kill them when unmounting
    this.tweenStroke = null;
    this.tweenScore = null;

    this.startAnimation = this.startAnimation.bind(this);
    this.displayContentReel = this.displayContentReel.bind(this);
    this.displayMarkers = this.displayMarkers.bind(this);
    this.killTweens = this.killTweens.bind(this);
    this.onSlideClick = this.onSlideClick.bind(this);
    this.onNewProps = this.onNewProps.bind(this);
  }

  componentDidMount() {
    this.onNewProps();
  }

  onNewProps() {
    let newState = {
      targetScore: this.props.slides[0].score,
      maxScore: this.props.slides[0].maxScore,
      currentSlideIndex: 0,
    };
    let stateChangeCallback = this.startAnimation;
    this.killTweens();
    if(!this.props.animate) {
      newState.currentCoefficientStroke = this.props.slides[0].score / this.props.slides[0].maxScore;
      newState.currentCoefficientNumber = newState.currentCoefficientStroke;
      stateChangeCallback = undefined;
    }

    this.setState(newState, stateChangeCallback);
  }

  componentWillUnmount() {
    this.killTweens();
  }

  componentDidUpdate(prevProps) {
    // this should probable be done in a smarter way
    if(JSON.stringify(this.props.slides) !== JSON.stringify(prevProps.slides)) {
      this.onNewProps();
    }
  }

  startAnimation() {
    this.killTweens();

    this.tweenScore = TweenMax.to(
      this,
      this.animationDuration,
      {
        valueTweenScore: this.state.targetScore / this.state.maxScore,
        ease: EaseTypes.Strong.easeOut,
        // we use the update handle from this tween to update both values instead
        // of doing it for both tweens in order to avoid re-rendering twice per cycle
        onUpdate: () => {
          this.setState({
            currentCoefficientNumber: this.valueTweenScore,
            currentCoefficientStroke: this.valueTweenStroke
          });
        }
      }
    );
    this.tweenStroke = TweenMax.to(
      this,
      this.animationDuration,
      {
        valueTweenStroke: this.state.targetScore / this.state.maxScore,
        ease: EaseTypes.Bounce.easeOut,
      }
    );
  }

  killTweens() {
    if(this.tweenStroke) {
      this.tweenStroke.kill();
    }
    if(this.tweenScore) {
      this.tweenScore.kill();
    }
  }

  displayContentReel() {
    return this.props.slides.map(slideData => {
      switch(slideData.type) {
        case 'score':
          return getScoreSlide(this.props, this.state);
        case 'offers':
          return getOffersSlide(this.props, this.state);
        case 'debt':
          return getDebtSlide(this.props, this.state);
        case 'balance':
          return getBalanceSlide(this.props, this.state);
        default:
          return null;
      }
    });
  }

  getRadius(size) {
    let radius = slideSizes.big;
    if(slideSizes.hasOwnProperty(size)) {
      radius = slideSizes[size];
    }
    return radius;
  }

  onSlideClick() {
    let nextSlide = this.state.currentSlideIndex + 1;
    if(this.state.currentSlideIndex >= this.props.slides.length - 1) {
      nextSlide = 0;
    }
    let stateChangeCallback;
    if(this.props.animate) {
      stateChangeCallback = this.startAnimation;
    }
    this.setState({
      currentSlideIndex: nextSlide,
      maxScore: this.props.slides[nextSlide].maxScore,
      targetScore: this.props.slides[nextSlide].score,
    }, stateChangeCallback);
  }

  displayMarkers() {
    if(this.props.slides.length <= 1) {
      return null;
    }
    return this.props.slides.map((slideData, index) => {
      let className = cn('marker');
      if(index === this.state.currentSlideIndex) {
        className += " " + cn('marker-selected');
      }
      return (
        <div key={`slide-${index}`} className={className}></div>
      )
    });
  }

  render() {
    if(
      !this.state.currentCoefficientStroke ||
      !this.state.currentCoefficientNumber ||
      !this.state.maxScore
    ) {
      return null;
    }

    const radius = this.getRadius(this.props.size);
    const strokeWidth = this.props.strokeWidth || 3;
    const spaceToEdge = 4;

    let angle = this.state.currentCoefficientStroke * 360;
    if(angle % 360 === 0) {
      angle = 359.9;
    }

    return (
      <div
        className={cn(null, `main-${this.props.size}`)}
        style={this.props.style}
        onClick={this.onSlideClick}
      >
        <div className={cn('bg')}></div>
        <div className={cn('border')}></div>
        <SVGArc
          strokeWidth={strokeWidth}
          radius={radius - strokeWidth - spaceToEdge}
          angle={angle}
          color={this.props.slides[this.state.currentSlideIndex].color}
        />
        <div className={cn('content')}>
          <div
            className={cn('reel')}
            style={{
              left: `-${this.state.currentSlideIndex * radius * 2}px`,
              width: `${this.props.slides.length * radius * 2}px`
            }}
          >
            {this.displayContentReel()}
          </div>
        </div>
        <div className={cn('marker-container')}>
          {this.displayMarkers()}
        </div>
      </div>
    )
  }
}
