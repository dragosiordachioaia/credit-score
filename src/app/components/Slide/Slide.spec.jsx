import React from 'react';
import renderer from 'react-test-renderer';
import TweenMax from 'gsap/TweenMax';

import * as helperMethods from './SlideHelper';

helperMethods.getScoreSlide = jest.fn().mockReturnValue('score-slide');
helperMethods.getOffersSlide = jest.fn().mockReturnValue('offers-slide');
helperMethods.getBalanceSlide = jest.fn().mockReturnValue('balance-slide');
helperMethods.getDebtSlide = jest.fn().mockReturnValue('debt-slide');

import Slide from './Slide';

const slideScore = {
  type: 'score',
  score: 100,
  maxScore: 700,
  color: '#a7ced1',
}

let component;

xdescribe('Slide component', () => {
  it('renders nothing right after being mounted', () => {
    const slides = [slideScore];
    const component = renderer.create(
      <Slide animate={true} size='big' slides={slides}/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })
  it('renders correctly when 1 slide', () => {
    const slides = [slideScore];

    const component = renderer.create(
      <Slide animate={true} size='big' slides={slides}/>
    );
    let instance = component.getInstance();
    instance.state = Object.assign(instance.state, {
      currentCoefficientStroke: 1,
      currentCoefficientNumber: 1,
      maxScore: 700,
    });
    let renderResult = instance.render();
    expect(renderResult).toMatchSnapshot();
  });
  it('kills tweens on new props', () => {
    const slides = [slideScore];
    const component = renderer.create(
      <Slide animate={true} size='big' slides={slides}/>
    );
    let instance = component.getInstance();
    instance.killTweens = jest.fn();
    instance.onNewProps();
    expect(instance.killTweens).toBeCalled();
  });
  it('calls setState on new props', () => {
    const slides = [slideScore];
    const component = renderer.create(
      <Slide animate={true} size='big' slides={slides}/>
    );
    let instance = component.getInstance();
    instance.setState = jest.fn();
    instance.onNewProps();
    expect(instance.setState).toBeCalled();
  });
  it('calls onNewProps in componentDidMount', () => {
    const slides = [slideScore];
    const component = renderer.create(
      <Slide animate={true} size='big' slides={slides}/>
    );
    let instance = component.getInstance();
    instance.onNewProps = jest.fn();
    instance.componentDidMount();
    expect(instance.onNewProps).toBeCalled();
  });
  it('kills tweens when unmounted', () => {
    const slides = [slideScore];
    const component = renderer.create(
      <Slide animate={true} size='big' slides={slides}/>
    );
    let instance = component.getInstance();
    instance.killTweens = jest.fn();
    instance.componentWillUnmount();
    expect(instance.killTweens).toBeCalled();
  });
  it('calls onNewProps if props change', () => {
    const oldProps = {
      size: 'big',
      slides: [slideScore],
    }
    const newProps = {
      size: 'big',
      slides: [slideScore, slideScore],
    }
    const component = renderer.create(
      <Slide {...oldProps}/>
    );
    let instance = component.getInstance();
    instance.onNewProps = jest.fn();
    instance.props = newProps;

    instance.componentDidUpdate(oldProps);
    expect(instance.onNewProps).toBeCalled();
  });
  it('does not call onNewProps if props do not change', () => {
    const oldProps = {
      size: 'big',
      slides: [slideScore],
    }
    const newProps = {
      size: 'big',
      slides: [slideScore],
    }
    const component = renderer.create(
      <Slide {...oldProps}/>
    );
    let instance = component.getInstance();
    instance.onNewProps = jest.fn();
    instance.props = newProps;

    instance.componentDidUpdate(oldProps);
    expect(instance.onNewProps).not.toBeCalled();
  });
  it('kills tweens on startAnimation', () => {
    const slides = [slideScore];
    const component = renderer.create(
      <Slide animate={true} size='big' slides={slides}/>
    );
    let instance = component.getInstance();
    instance.killTweens = jest.fn();
    instance.startAnimation();
    expect(instance.killTweens).toBeCalled();
  });
  it('kills tweens on startAnimation', () => {
    const slides = [slideScore];
    TweenMax.to = jest.fn();
    const component = renderer.create(
      <Slide animate={true} size='big' slides={slides}/>
    );
    let instance = component.getInstance();
    instance.startAnimation();
    expect(TweenMax.to).toBeCalled();
  });
  it('correctly displays score slide', () => {
    const slides = [slideScore];
    const component = renderer.create(
      <Slide animate={true} size='big' slides={slides}/>
    );
    let instance = component.getInstance();
    let result = instance.displayContentReel();
    expect(result).toMatchSnapshot();
  });
  it('correctly displays offers slide', () => {
    const slide = Object.assign({}, slideScore, {type: 'offers'});
    const component = renderer.create(
      <Slide animate={true} size='big' slides={[slide]}/>
    );
    let instance = component.getInstance();
    let result = instance.displayContentReel();
    expect(result).toMatchSnapshot();
  });
  it('correctly displays debt slide', () => {
    const slide = Object.assign({}, slideScore, {type: 'debt'});
    const component = renderer.create(
      <Slide animate={true} size='big' slides={[slide]}/>
    );
    let instance = component.getInstance();
    let result = instance.displayContentReel();
    expect(result).toMatchSnapshot();
  });
  it('correctly displays balance slide', () => {
    const slide = Object.assign({}, slideScore, {type: 'balance'});
    const component = renderer.create(
      <Slide animate={true} size='big' slides={[slide]}/>
    );
    let instance = component.getInstance();
    let result = instance.displayContentReel();
    expect(result).toMatchSnapshot();
  });
  it('correctly displays nothing for unknown slide type', () => {
    const slide = Object.assign({}, slideScore, {type: 'unknown_slide'});
    const component = renderer.create(
      <Slide animate={true} size='big' slides={[slide]}/>
    );
    let instance = component.getInstance();
    let result = instance.displayContentReel();
    expect(result).toMatchSnapshot();
  });
  it('properly returns radius', () => {
    const component = renderer.create(
      <Slide animate={true} size='big' slides={[slideScore]}/>
    );
    let instance = component.getInstance();
    let result;

    result = instance.getRadius('big');
    expect(result).toMatchSnapshot();
    result = instance.getRadius('medium');
    expect(result).toMatchSnapshot();
    result = instance.getRadius('small');
    expect(result).toMatchSnapshot();
    result = instance.getRadius('superbig');
    expect(result).toMatchSnapshot();
  });
  it('correctly skips markers for 1 slide', () => {
    const component = renderer.create(
      <Slide animate={true} size='big' slides={[slideScore]}/>
    );
    let instance = component.getInstance();
    let result = instance.displayMarkers();
    expect(result).toMatchSnapshot();
  });
  it('correctly skips markers for no slides', () => {
    const component = renderer.create(
      <Slide animate={true} size='big' slides={[slideScore]}/>
    );
    let instance = component.getInstance();
    let result = instance.displayMarkers();
    expect(result).toMatchSnapshot();
  });
  it('correctly displays markers for 2 slides', () => {
    const component = renderer.create(
      <Slide animate={true} size='big' slides={[slideScore, slideScore]}/>
    );
    let instance = component.getInstance();
    let result = instance.displayMarkers();
    expect(result).toMatchSnapshot();
  });
});
