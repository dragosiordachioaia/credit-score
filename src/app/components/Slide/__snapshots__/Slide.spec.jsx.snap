// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slide component correctly displays balance slide 1`] = `
Array [
  "balance-slide",
]
`;

exports[`Slide component correctly displays debt slide 1`] = `
Array [
  "debt-slide",
]
`;

exports[`Slide component correctly displays markers for 2 slides 1`] = `
Array [
  <div
    className="slide__marker slide__marker-selected"
  />,
  <div
    className="slide__marker"
  />,
]
`;

exports[`Slide component correctly displays nothing for unknown slide type 1`] = `
Array [
  null,
]
`;

exports[`Slide component correctly displays offers slide 1`] = `
Array [
  "offers-slide",
]
`;

exports[`Slide component correctly displays score slide 1`] = `
Array [
  "score-slide",
]
`;

exports[`Slide component correctly skips markers for 1 slide 1`] = `null`;

exports[`Slide component correctly skips markers for no slides 1`] = `null`;

exports[`Slide component properly returns radius 1`] = `150`;

exports[`Slide component properly returns radius 2`] = `120`;

exports[`Slide component properly returns radius 3`] = `85`;

exports[`Slide component properly returns radius 4`] = `150`;

exports[`Slide component renders correctly when 1 slide 1`] = `
<div
  className="slide slide--main-big"
  onClick={[Function]}
  style={undefined}
>
  <div
    className="slide__bg"
  />
  <div
    className="slide__border"
  />
  <SVGArc
    angle={359.9}
    color="#a7ced1"
    radius={143}
    strokeWidth={3}
  />
  <div
    className="slide__content"
  >
    <div
      className="slide__reel"
      style={
        Object {
          "left": "-0px",
          "width": "300px",
        }
      }
    >
      score-slide
    </div>
  </div>
  <div
    className="slide__marker-container"
  />
</div>
`;

exports[`Slide component renders nothing right after being mounted 1`] = `null`;
